File: public/java/009-longest-substring-without-repeating.java
/*
 * LeetCode Problem #3: Longest Substring Without Repeating Characters
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/longest-substring-without-repeating-characters/
 */
import java.util.*;

class Solution {

    // APPROACH 1: Brute Force | O(n³) time | O(k) space
    // EXPLAIN: Check every substring and verify it has all unique characters using a set.
    // WHEN: Only for tiny strings; demonstrates the naive O(n³) baseline.
    public int lengthOfLongestSubstring_BruteForce(String s) {
        int maxLen = 0;
        int n = s.length();
        for (int i = 0; i < n; i++) {
            Set<Character> seen = new HashSet<>();
            for (int j = i; j < n; j++) {
                char c = s.charAt(j);
                if (seen.contains(c)) break;
                seen.add(c);
                maxLen = Math.max(maxLen, j - i + 1);
            }
        }
        return maxLen;
    }

    // APPROACH 2: Sliding Window with Set | O(n) time | O(k) space
    // EXPLAIN: Maintain a set of window characters; shrink from the left when a duplicate is encountered.
    // WHEN: Easy to explain and implement; good introduction to the sliding window pattern.
    public int lengthOfLongestSubstring_Set(String s) {
        Set<Character> window = new HashSet<>();
        int maxLen = 0, left = 0;
        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            while (window.contains(c)) {
                window.remove(s.charAt(left));
                left++;
            }
            window.add(c);
            maxLen = Math.max(maxLen, right - left + 1);
        }
        return maxLen;
    }

    // APPROACH 3: Sliding Window with HashMap | O(n) time | O(k) space
    // EXPLAIN: Maintain a window [left, right]; on duplicate, jump left past the previous occurrence.
    // WHEN: Optimal — single pass, handles all character sets via a map from char to last index.
    public int lengthOfLongestSubstring_SlidingWindow(String s) {
        Map<Character, Integer> lastIndex = new HashMap<>();
        int maxLen = 0;
        int left = 0;
        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            if (lastIndex.containsKey(c) && lastIndex.get(c) >= left) {
                left = lastIndex.get(c) + 1;
            }
            lastIndex.put(c, right);
            maxLen = Math.max(maxLen, right - left + 1);
        }
        return maxLen;
    }

    // APPROACH 4: Array Instead of HashMap | O(n) time | O(1) space
    // EXPLAIN: Use a 128-element int array indexed by ASCII value instead of a HashMap.
    // WHEN: When input is guaranteed ASCII — array access is faster than hash map operations.
    public int lengthOfLongestSubstring_Array(String s) {
        int[] charIndex = new int[128];
        Arrays.fill(charIndex, -1);
        int maxLen = 0, left = 0;
        for (int right = 0; right < s.length(); right++) {
            int code = s.charAt(right);
            if (charIndex[code] >= left) {
                left = charIndex[code] + 1;
            }
            charIndex[code] = right;
            maxLen = Math.max(maxLen, right - left + 1);
        }
        return maxLen;
    }
}

// Made with Bob

File: public/python/009-longest-substring-without-repeating.py
# LeetCode Problem #3: Longest Substring Without Repeating Characters
# Difficulty: Medium
# Link: https://leetcode.com/problems/longest-substring-without-repeating-characters/

# ─────────────────────────────────────────────
# APPROACH 1: Brute Force | O(n³) time | O(min(m,n)) space
# EXPLAIN: Generate all substrings with two loops; check each for duplicates using a set.
# WHEN: Tiny inputs only — establishes the baseline before sliding window solutions.

def length_of_longest_brute(s: str) -> int:
    n = len(s)
    best = 0
    for i in range(n):
        for j in range(i + 1, n + 1):
            if len(set(s[i:j])) == j - i:   # all unique
                best = max(best, j - i)
    return best


# ─────────────────────────────────────────────
# APPROACH 2: Sliding Window with Set | O(n) time | O(min(m,n)) space
# EXPLAIN: Maintain a set of characters in the current window; shrink from the left when a repeat is found.
# WHEN: Clean and intuitive — great for explaining the sliding window pattern in interviews.

def length_of_longest_sliding_window(s: str) -> int:
    char_set: set[str] = set()
    lo = best = 0
    for hi, ch in enumerate(s):
        while ch in char_set:
            char_set.remove(s[lo])
            lo += 1
        char_set.add(ch)
        best = max(best, hi - lo + 1)
    return best


# ─────────────────────────────────────────────
# APPROACH 3: Sliding Window with HashMap | O(n) time | O(min(m,n)) space
# EXPLAIN: Store the last-seen index of each character; jump the left pointer directly past the duplicate.
# WHEN: Slightly fewer iterations in practice — preferred when you want to minimise pointer movements.

def length_of_longest_optimised(s: str) -> int:
    last_seen: dict[str, int] = {}
    lo = best = 0
    for hi, ch in enumerate(s):
        if ch in last_seen and last_seen[ch] >= lo:
            lo = last_seen[ch] + 1            # jump past the previous occurrence
        last_seen[ch] = hi
        best = max(best, hi - lo + 1)
    return best


# ─────────────────────────────────────────────
# APPROACH 4: Array Instead of HashMap | O(n) time | O(1) space
# EXPLAIN: Use a 128-element integer array indexed by ASCII code instead of a hash map.
# WHEN: When input is guaranteed ASCII — array lookups avoid hash overhead for maximum speed.

def length_of_longest_array(s: str) -> int:
    char_index = [-1] * 128   # index of last seen position for each ASCII char
    lo = best = 0
    for hi, ch in enumerate(s):
        code = ord(ch)
        if char_index[code] >= lo:
            lo = char_index[code] + 1
        char_index[code] = hi
        best = max(best, hi - lo + 1)
    return best


# ─────────────────────────────────────────────
if __name__ == "__main__":
    cases = [
        ("abcabcbb", 3),
        ("bbbbb",    1),
        ("pwwkew",   3),
        ("",         0),
    ]
    for s, expected in cases:
        assert length_of_longest_brute(s)           == expected
        assert length_of_longest_sliding_window(s)  == expected
        assert length_of_longest_optimised(s)       == expected
        assert length_of_longest_array(s)           == expected
    print("All tests passed.")

# Made with Bob
